lib/HLSL/HLOperations.cpp

```diff
@@ -509,9 +509,15 @@ Function *GetOrCreateHLFunction(Module &M, FunctionType *funcTy,
   }
   else {
     mangledNameStr << GetHLFullName(group, opcode);
-    // Need to add wave sensitivity to name to prevent clashes with non-wave intrinsic
-    if(attribs.hasAttribute(AttributeSet::FunctionIndex, HLWaveSensitive))
-        mangledNameStr << "wave";
+    // Need to add attributes to name to prevent clashes with intrinsics with
+    // different attributes
+    if (attribs.hasAttribute(AttributeSet::FunctionIndex, Attribute::ReadNone))
+      mangledNameStr << "rn";
+    else if (attribs.hasAttribute(AttributeSet::FunctionIndex,
+                                  Attribute::ReadOnly))
+      mangledNameStr << "ro";
+    if (attribs.hasAttribute(AttributeSet::FunctionIndex, HLWaveSensitive))
+      mangledNameStr << "wave";
```

Comment:
What about other attributes? I see we ignore `nounwind` based on the tests, but maybe we should do that explicitly and warn or error or something if we see something odd, otherwise this is just going to pop up again.

Reply:
The only relevant attributes for this issue are those defined in gen_intrin_main

Reply:
`nounwind` is always added to these intrinsic functions, so isn't presumed to be a dimension for overloading.  Since these are just for mangling the built-in intrinsics, there is an assumption that the attributes used are limited to a certain set.

However, I do see that we have special cases where we add attributes under `SetHLFunctionAttribute()`, including potentially `NoDuplicate`.  That isn't accounted for in this code.

Ideally, one place would define the correct attributes on the original functions, and the mangling would follow, but this code isn't written to be able to do that, since special cases for attributes are applied after the mangling was defined, and the function looked up.  I think combining the attribute adding and attribute mangling code into the same function would prevent them from getting out of sync so easily, at least.

Some of these attributes are definable in the intrinsic table input generated from `gen_intrin_main.txt` (ro/rn/wave), and I think they all should be (at least the ones that could be different).  There also should be use of `ArgMemOnly` in SM 6.8 branch, and another old PR I had up here changed the way the attributes are defined in the intrinsic tables to allow for easier expansion of attributes as needed.  But some of these functions don't come from the intrinsic tables, and are generated elsewhere as built-in operators or intermediate operations, and some of those are special-cased in `SetHLFunctionAttribute()`.